rm-datasource/src/main/java/io/seata/rm/RMHandlerAT.java

```diff
@@ -40,43 +43,93 @@
+            deleteRows = deleteUndoLog(manager, conn, division);
+        } while (deleteRows == LIMIT_ROWS);
+
         try {
-            conn = dataSourceProxy.getPlainConnection();
-            int deleteRows = 0;
-            do {
-                try {
-                    deleteRows = UndoLogManagerFactory.getUndoLogManager(dataSourceProxy.getDbType())
-                            .deleteUndoLogByLogCreated(logCreatedSave, LIMIT_ROWS, conn);
-                    if (deleteRows > 0 && !conn.getAutoCommit()) {
-                        conn.commit();
-                    }
-                } catch (SQLException exx) {
-                    if (deleteRows > 0 && !conn.getAutoCommit()) {
-                        conn.rollback();
-                    }
-                    throw exx;
-                }
-            } while (deleteRows == LIMIT_ROWS);
-        } catch (Exception e) {
-            LOGGER.error("Failed to delete expired undo_log, error:{}", e.getMessage(), e);
-        } finally {
-            if (conn != null) {
-                try {
-                    conn.close();
-                } catch (SQLException closeEx) {
-                    LOGGER.warn("Failed to close JDBC resource while deleting undo_log ", closeEx);
+            conn.close();
+        } catch (SQLException closeEx) {
+            LOGGER.warn("Failed to close JDBC resource after deleting undo_log ", closeEx);
+        }
+    }
+
+    boolean checkUndoLogTableExist(DataSourceProxy dataSourceProxy) {
+        Connection connection = getConnection(dataSourceProxy);
+        if (connection == null) {
+            return false;
+        }
+        UndoLogManager manager = getUndoLogManager(dataSourceProxy);
+        boolean isExisted = manager.hasUndoLogTable(connection);
+        try {
+            connection.close();
+        } catch (SQLException e) {
+            String resourceId = dataSourceProxy.getResourceId();
+            LOGGER.warn("Failed to close JDBC resource for {}", resourceId);
+        }
+        return isExisted;
+    }
+
+    Connection getConnection(DataSourceProxy dataSourceProxy) {
+        try {
+            return dataSourceProxy.getPlainConnection();
+        } catch (SQLException e) {
+            String resourceId = dataSourceProxy.getResourceId();
+            LOGGER.error("Failed to get connection for {}", resourceId, e);
+            return null;
+        }
+    }
+
+    UndoLogManager getUndoLogManager(DataSourceProxy dataSourceProxy) {
+        return UndoLogManagerFactory.getUndoLogManager(dataSourceProxy.getDbType());
+    }
+
+    int deleteUndoLog(UndoLogManager manager, Connection conn, Date division) {
+        try {
+            int deleteRows = manager.deleteUndoLogByLogCreated(division, LIMIT_ROWS, conn);
+            if (!conn.getAutoCommit()) {
+                conn.commit();
+            }
+            return deleteRows;
+        } catch (SQLException e) {
+            LOGGER.error("Failed to delete expired undo_log", e);
```

Comment:
LOGGER.error("Failed to delete expired undo_log", e.getMessage(),e);

Reply:
message will be printed automatically.